executor/join_result_generators.go

```diff
@@ -133,15 +161,22 @@ type semiJoinResultGenerator struct {
 	baseJoinResultGenerator
 }
 
-// emit implements joinResultGenerator interface.
-func (outputer *semiJoinResultGenerator) emit(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) error {
-	if inners == nil || inners.Len() == 0 {
-		return nil
+func (outputer *semiJoinResultGenerator) tryToMatch(outer chunk.Row, inners chunk.Iterator, chk *chunk.Chunk) (matched bool, err error) {
+	if inners.Len() == 0 {
+		return false, nil
 	}
-	defer inners.ReachEnd()
-	if len(outputer.conditions) == 0 {
+
+	defer func() {
+		if !(err == nil && matched) {
+			return
+		}
+		// here we handle the matched outer.
 		chk.AppendPartialRow(0, outer)
```

Comment:
I think do the main work in defer function is too tricky.
How about just
```
defer func() {
    if matched {
        innters.ReachEnd()
    }
}
```

And `AppendPartialRow` before `return true, nil`.

Reply:
Actually, `innters.ReachEnd()` should be called together when we have a matched inner row and do the join work. So I chose the `defer` style to reduce the code duplication.

Reply:
I think we should not use `defer` to reducing code duplication.
It splits the execution flow, makes the code harder to read.